### PR TITLE
Adjust get single table data source name mapping.

### LINKED ISSUE #15191: Table changed not refresh SingleTableRule.
### Which version of ShardingSphere did you use?

`master`

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

`ShardingSphere-Proxy`

### Expected behavior

Refresh SingleTableRule when table meta data changed.

### Actual behavior

No refresh.

image

### PR DESCRIPTION
**Pull request by zjcnb**
Fixes #15191, #15207

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/rule/identifier/type/MutableDataNodeRule.java [hunk 1]
```diff
@@ -42,9 +43,10 @@
...
     /**
-     * Get data source names.
+     * Find single data node by table name.
      *
-     * @return Get data source names
+     * @param tableName table name
+     * @return single table data node
      */
-    Collection<String> getDataSourceNames();
+    Optional<DataNode> findSingleTableDataNodeByTableName(String tableName);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/rule/identifier/type/MutableDataNodeRule.java:51]: @zjcnb Why not change this method to findSingleTableDataNode directly?

File: shardingsphere-mode/shardingsphere-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/ContextManager.java [hunk 2]
```diff
@@ -251,9 +251,7 @@ public void alterSchema(final String schemaName, final ShardingSphereSchema sche
      */
     public void alterSchema(final String schemaName, final TableMetaData changedTableMetaData, final String deletedTable) {
         ShardingSphereMetaData metaData = metaDataContexts.getMetaData(schemaName);
-        if (!containsInDataNodeContainedRule(changedTableMetaData.getName(), metaData)) {
-            metaData.getRuleMetaData().findRules(MutableDataNodeRule.class).forEach(each -> each.put(changedTableMetaData.getName(), each.getDataSourceNames().iterator().next()));
-        }
+        singleTableDataSourceMappingRelation(metaData, changedTableMetaData);
```
Comment (strongduanmu) [inline shardingsphere-mode/shardingsphere-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/ContextManager.java:254]: Can we use another meaningful method name? Just like 'alterSingleTableDataNodes'?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/rule/identifier/type/MutableDataNodeRule.java (hunk 1 above)
Comment (zjcnb) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/rule/identifier/type/MutableDataNodeRule.java:51]: Yes, I will adjust it.

File: shardingsphere-mode/shardingsphere-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/ContextManager.java (hunk 2 above)
Comment (zjcnb) [reply shardingsphere-mode/shardingsphere-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/ContextManager.java:254]: Yes, it' s a good idea.

**Review by strongduanmu — APPROVED** [review]

**Pull request merged**